clash-lib/src/Data/Aeson/Extra.hs

```diff
@@ -4,47 +4,122 @@
-replaceCommonEscapes :: Text -> Text
-replaceCommonEscapes = ( T.replace (pack "\\n") (pack "\n") ) .
-                       ( T.replace (pack "\\\\") (pack "\\") ) .
-                       ( T.replace (pack "\\\"") (pack "\"") )
+-- | See 'toSpecNewlines'. A line map maps "virtual" lines to a range of
+-- "real" lines. E.g., a map of {0: (0, 3), 1: (4, 5)} would mean that line
+-- 0 in the virtual JSON (i.e., the one with newlines replaced) file map to
+-- lines 0 up to and including 3 in the original user-supplied one.
+type LineMap = IntMap (Int, Int)
+
+-- | Aeson versions <1.5.0 accept unescaped newlines in JSON strings. This is in
+-- violation of RFC 7159. Aeson 1.5.0 fixes this bug. Unfortunately, "Clash
+-- JSON" files rely on the old behavior. This function replaces newlines (in
+-- stings) with their escaped variants.
+toSpecNewlines
+  :: ByteString
+  -> Either UnicodeException (LineMap, ByteString)
+toSpecNewlines bs = do
+ s0 <- LT.unpack <$> LT.decodeUtf8' bs
+ Right ( toLineMap (go2 0 False s0)
+       , LT.encodeUtf8 (LT.pack (go False s0)))
+ where
+  -- replace newlines with escaped ones
+  go :: Bool -> String -> String
+  go _ [] = []
+  go True ('\n':rest) = '\\' : 'n' : go True rest
+  go True ('\r':rest) = '\\' : 'r' : go True rest
+  go True ('\t':rest) = '\\' : 't' : go True rest
+  go inString ('\\':r:rest) = '\\' : r : go inString rest
+  go inString ('"':rest) = '"' : go (not inString) rest
+  go inString (r:rest) = r : go inString rest
+
+  -- Calculate real:virtual mapping
+  go2
+    -- virtual line counter.
+    :: Int
+    -- Processing a JSON string?
+    -> Bool
+    -- String left to process
+    -> String
+    -- Virtual line numbers. [0, 1, 1, 2, 2, ..] would mean:
+    --
+    --  real | virtual
+    --  --------------
+    --     0 |       0
+    --     1 |       1
+    --     2 |       1
+    --     3 |       2
+    --     4 |       2
+    --   ... |     ...
+    --
+    -> [Int]
+  go2 n _        [] = [n]
+  go2 n True     ('\n':rest) = n : go2 n True rest
+  go2 n False    ('\n':rest) = n : go2 (succ n) False rest
+  go2 n inString ('\\':_:rest) = go2 n inString rest
+  go2 n inString ('"':rest) = go2 n (not inString) rest
+  go2 n inString (_:rest) = go2 n inString rest
 
-genLineErr' :: [Text] -> (Int, Int) -> Int -> Text
-genLineErr' allLines range errorLineN = T.unlines [ T.concat [ if i == errorLineN then pack ">> " else  pack "   "
-                                                             , pack $ show i
-                                                             , pack ". "
-                                                             , allLines !! i
-                                                             ] | i <- Ix.range range]
+  toLineMap :: [Int] -> LineMap
+  toLineMap virtuals =
+      IntMap.fromList
+    $ map (second (\reals -> (minimum reals, maximum reals)))
+    $ map (\(unzip -> (virts, reals)) -> (head virts, reals))
+    $ groupBy (\a b -> fst a == fst b)
```

Comment:
Few suggestions:

  * `\(unzip -> (virts, reals)) -> (head virts, reals)` -> `first head. unzip`
  * `groupBy (\a b -> fst a == fst b)` -> `groupOn fst` (from `Data.List.Extra`)

Reply:
ooh `groupOn`, never seen that before

Reply:
If you really want to invoke your inner perl programmer, you can also just do one `map` instead of two and make it a pointfree one-liner:

```haskell
IntMap.fromList . fmap (bimap head (minimum &&& maximum) . unzip) . groupOn fst
```

Reply:
I'm going to have to respectfully decline, I don't think that'd make it easier to read! (But I guess that's why you alluded to it being Perl-like like ;))